x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/fulltext/QueryStringFunctionTests.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.esql.expression.function.fulltext;
+
+import com.carrotsearch.randomizedtesting.annotations.Name;
+import com.carrotsearch.randomizedtesting.annotations.ParametersFactory;
+
+import org.apache.lucene.util.BytesRef;
+import org.elasticsearch.xpack.esql.core.expression.Expression;
+import org.elasticsearch.xpack.esql.core.tree.Source;
+import org.elasticsearch.xpack.esql.core.type.DataType;
+import org.elasticsearch.xpack.esql.expression.function.AbstractFunctionTestCase;
+import org.elasticsearch.xpack.esql.expression.function.TestCaseSupplier;
+import org.hamcrest.Matcher;
+
+import java.util.Arrays;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.function.Supplier;
+
+import static org.hamcrest.Matchers.equalTo;
+
+public class QueryStringFunctionTests extends AbstractFunctionTestCase {
```

Comment:
Question - do we need to generate docs for functions that are just used in snapshot builds? Is there a way to prevent these to end up in docs until we have them ready for our users?

Reply:
I dug deeper on this - as long as the function is registered in the SnapshotRepository, no docs are generated for it. We need to make sure docs are generated when we remove QSTR from the snapshot build.

Reply:
We'll still want that documentation. As of https://github.com/elastic/elasticsearch/pull/112792 the plan was to provide the docs for snapshot functions, but to include the `preview` flag...

The reason is that we want the editor to be able to validate the function, even if it is in development. To do that, we need the function signature.

We will, however, not _suggest_ the function if `preview` is set to `true`.

Reply:
@drewdaemon @fang-xing-esql , currently snapshot functions do not generate documentation of any kind.

We don't want snapshot functions to be included in our ES docs, correct? IIUC, that would mean [changing the docs generation process](https://github.com/elastic/elasticsearch/blob/ef3a5a138530c43b8e6cd94585d13b853b819c57/x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/AbstractFunctionTestCase.java#L864) so Kibana docs are generated for all functions (including snapshot and preview). ES docs would keep being generated just for non-snapshot ones (including preview).

Is this correct? Is there work already in progress for doing this change, or should we do that on a separate PR?

Reply:
Snapshot functions are relatively new to us, thank you for putting a lot of effort on this @carlosdelest @ioanatia, it is in pretty good shape, and it will be a good example of snapshot functions moving forward. The only thing left that I can think of for now is docs.

There are a few steps below that we can do to make a docs available(if we decide to do it in this PR or as a follow up), if it can help Kibana to provide a better user experience, I'll rely on @drewdaemon 's input, as I have very little knowledge about how Kibana handles snapshot functions.

1. Add annotation to `QueryStringFunctionTests`, so that `AbstractFunctionTestCase` can recognize it an a registered function
```
@FunctionName("qstr")
public class QueryStringFunctionTests extends AbstractFunctionTestCase {
```

2. Make `AbstractFunctionTestCase` generate docs for snapshot functions.
```
    private static FunctionDefinition definition(String name) {
        EsqlFunctionRegistry registry = new EsqlFunctionRegistry().snapshotRegistry();
        if (registry.functionExists(name)) {
            return registry.resolveFunction(name);
        }
        return null;
    }
```

3. Include `ifeval::["{release-state}"=="unreleased"]` in the docs where this function will be linked in. There are a couple of examples for snapshot commands(inlinestats, lookup) here
```
ifeval::["{release-state}"=="unreleased"]
* experimental:[] <<esql-inlinestats-by>>
endif::[]
* <<esql-keep>>
* <<esql-limit>>
ifeval::["{release-state}"=="unreleased"]
* experimental:[] <<esql-lookup>>
endif::[]
```

ES|QL functions are in `esql-functions-operators.asciidoc`, `QSTR` can be linked under it, but don't forget to wrap is around with `ifeval::["{release-state}"=="unreleased"]` if it is still under snapshot.

Reply:
Thanks @fang-xing-esql ! I've generated the docs, and it seems that functions docs need to be explicitly linked from the [functions top level docs](https://github.com/elastic/elasticsearch/blob/b26d7d3e1489ffb35283603599a83cbb4762dcfe/docs/reference/esql/esql-functions-operators.asciidoc?plain=1#L86) - so we should be safe as the docs are there but not linked to the top-level pages.

I've merged main, resolved conflicts, and committed the generated docs that include Kibana docs.

LMK if this works! 👍

Reply:
Looks great @carlosdelest thank you for your patience.

> We have two other functions available only in snapshot builds - rate and categorize - none of them have Kibana docs.

@ioanatia thank you for pointing this out!

This thread has highlighted that we confirm the user experience we want for snapshot functions in the editor and let that drive the documentation requirements. I will start that discussion outside of this PR.

Thanks all!